File: apps/bff/src/modules/task/router.ts
import type { ProjectStore, SyncReadStore, TaskStore } from '@bff/modules/task/ports';
import { DOMAIN_ERROR_CODES, DomainError } from '@bff/platform/errors';
import type { SyncBroker } from '@bff/platform/sync-broker';
import { protectedProcedure, router } from '@bff/trpc/init';
import type { ProjectDetailResponse, ProjectListResponse, SyncChangesResponse, Task, TaskRangeResponse } from '@tooday/shared';
import {
  createProjectRequestSchema,
  createTaskRequestSchema,
  projectDetailRequestSchema,
  syncChangesRequestSchema,
  taskIdRequestSchema,
  taskRangeRequestSchema,
  updateTaskRequestSchema,
} from '@tooday/shared';

export interface TaskRouterDeps {
  tasks: TaskStore;
  projects: ProjectStore;
  syncReads: SyncReadStore;
  sync: SyncBroker;
}

export function createTaskRouter({ tasks, projects, syncReads, sync }: TaskRouterDeps) {
  return router({
    /**
     * 메인(오늘) 화면 주간 창 데이터 + 동기화 커서 베이스라인.
     * 데이터와 커서는 스토어가 단일 스냅샷에서 함께 읽는다 — 라우터에서 낱개
     * 조회를 조합하면 스냅샷이 갈라져 커서가 못 본 변경을 지나칠 수 있다.
     */
    range: protectedProcedure.input(taskRangeRequestSchema).query(async ({ ctx, input }): Promise<TaskRangeResponse> => {
      return syncReads.range({ userId: ctx.userId, ...input });
    }),

    /** 태스크 상세 화면 — 단건 조회 */
    byId: protectedProcedure.input(taskIdRequestSchema).query(async ({ ctx, input }): Promise<{ task: Task }> => {
      const task = await tasks.findById({ userId: ctx.userId, id: input.id });
      if (!task) {
        throw new DomainError(DOMAIN_ERROR_CODES.TASK_NOT_FOUND);
      }
      return { task };
    }),

    create: protectedProcedure.input(createTaskRequestSchema).mutation(async ({ ctx, input }) => {
      if (input.projectId !== null) {
        const project = await projects.findById({ userId: ctx.userId, id: input.projectId });
        if (!project) {
          throw new DomainError(DOMAIN_ERROR_CODES.PROJECT_NOT_FOUND);
        }
      }
      const task = await tasks.create({ userId: ctx.userId, ...input });
      sync.notify(ctx.userId);
      return { task };
    }),

    /**
     * 의도 기반 부분 업데이트 — patch의 필드만 최신 행에 적용한다.
     * 같은 필드의 경합은 나중 의도가 이기고(LWW), 다른 필드끼리는 충돌하지 않는다.
     */
    update: protectedProcedure.input(updateTaskRequestSchema).mutation(async ({ ctx, input }) => {
      if (typeof input.patch.projectId === 'string') {
        const project = await projects.findById({ userId: ctx.userId, id: input.patch.projectId });
        if (!project) {
          throw new DomainError(DOMAIN_ERROR_CODES.PROJECT_NOT_FOUND);
        }
      }
      const task = await tasks.update({ userId: ctx.userId, ...input });
      if (!task) {
        throw new DomainError(DOMAIN_ERROR_CODES.TASK_NOT_FOUND);
      }
      sync.notify(ctx.userId);
      return { task };
    }),

    /** 소프트 삭제 — tombstone이 델타로 다른 기기에 전파된다 */
    delete: protectedProcedure.input(taskIdRequestSchema).mutation(async ({ ctx, input }) => {
      const removed = await tasks.remove({ userId: ctx.userId, id: input.id });
      if (!removed) {
        throw new DomainError(DOMAIN_ERROR_CODES.TASK_NOT_FOUND);
      }
      sync.notify(ctx.userId);
      return { id: input.id };
    }),

    /** 델타 동기화 — 커서 이후의 변경 전부 (tombstone 포함), 단일 스냅샷 (range 주석 참조) */
    changes: protectedProcedure.input(syncChangesRequestSchema).query(async ({ ctx, input }): Promise<SyncChangesResponse> => {
      return syncReads.changes({ userId: ctx.userId, cursor: input.cursor });
    }),

    createProject: protectedProcedure.input(createProjectRequestSchema).mutation(async ({ ctx, input }) => {
      const project = await projects.create({ userId: ctx.userId, ...input });
      sync.notify(ctx.userId);
      return { project };
    }),

    /** 프로젝트 목록 화면 — 프로젝트 라벨에 진행률(완료/전체)을 붙여 내려준다 */
    projects: protectedProcedure.query(async ({ ctx }): Promise<ProjectListResponse> => {
      const [projectList, counts] = await Promise.all([projects.listByUser(ctx.userId), tasks.countsByProject(ctx.userId)]);
      const countById = new Map(counts.map((count) => [count.projectId, count]));
      return {
        projects: projectList.map((project) => {
          const count = countById.get(project.id);
          return { ...project, totalCount: count?.total ?? 0, doneCount: count?.done ?? 0 };
        }),
      };
    }),

    /** 프로젝트 상세(보드) — 프로젝트 라벨과 그 프로젝트의 태스크 전부 */
    project: protectedProcedure
      .input(projectDetailRequestSchema)
      .query(async ({ ctx, input }): Promise<ProjectDetailResponse> => {
        const project = await projects.findById({ userId: ctx.userId, id: input.projectId });
        if (!project) {
          throw new DomainError(DOMAIN_ERROR_CODES.PROJECT_NOT_FOUND);
        }
        const taskList = await tasks.listByProject({ userId: ctx.userId, projectId: input.projectId });
        return { project, tasks: taskList };
      }),
  });
}

File: rust/crates/bff/src/modules/task/router.rs
use std::collections::HashMap;
use std::sync::Arc;

use serde_json::Value;
use tooday_shared::{
    CreateProjectRequest, CreateTaskRequest, DeletedTaskResponse, Patch, ProjectDetailRequest,
    ProjectDetailResponse, ProjectListResponse, ProjectResponse, ProjectSummary,
    SyncChangesRequest, SyncChangesResponse, TaskIdRequest, TaskRangeRequest, TaskRangeResponse,
    TaskResponse, UpdateTaskRequest,
};

use crate::modules::task::ports::{
    CreateProjectInput, CreateTaskInput, ListChangesInput, ListTasksByProjectInput,
    ListTasksRangeInput, ProjectStore, TaskRefInput, TaskStore, UpdateTaskInput,
};
use crate::platform::errors::{DomainError, DomainErrorCode};
use crate::platform::sync_broker::SyncBroker;
use crate::trpc::context::TrpcContext;
use crate::trpc::init::{
    output, parse_input, unauthenticated, AuthLevel, Procedure, ProcedureKind, TrpcError,
};

pub const PROCEDURES: &[Procedure] = &[
    Procedure { name: "range", kind: ProcedureKind::Query, auth: AuthLevel::Protected },
    Procedure { name: "byId", kind: ProcedureKind::Query, auth: AuthLevel::Protected },
    Procedure { name: "create", kind: ProcedureKind::Mutation, auth: AuthLevel::Protected },
    Procedure { name: "update", kind: ProcedureKind::Mutation, auth: AuthLevel::Protected },
    Procedure { name: "delete", kind: ProcedureKind::Mutation, auth: AuthLevel::Protected },
    Procedure { name: "changes", kind: ProcedureKind::Query, auth: AuthLevel::Protected },
    Procedure { name: "createProject", kind: ProcedureKind::Mutation, auth: AuthLevel::Protected },
    Procedure { name: "projects", kind: ProcedureKind::Query, auth: AuthLevel::Protected },
    Procedure { name: "project", kind: ProcedureKind::Query, auth: AuthLevel::Protected },
];

pub struct TaskRouterDeps {
    pub tasks: Arc<dyn TaskStore>,
    pub projects: Arc<dyn ProjectStore>,
    pub sync: Arc<dyn SyncBroker>,
}

fn project_not_found() -> TrpcError {
    DomainError::new(DomainErrorCode::ProjectNotFound).into()
}

fn task_not_found() -> TrpcError {
    DomainError::new(DomainErrorCode::TaskNotFound).into()
}

pub async fn call(
    deps: &TaskRouterDeps,
    ctx: &mut TrpcContext,
    name: &str,
    input: Option<Value>,
) -> Result<Value, TrpcError> {
    let user_id = ctx.user_id.clone().ok_or_else(unauthenticated)?;

    match name {
        // 메인(오늘) 화면 주간 창 데이터 + 동기화 커서
        "range" => {
            let request: TaskRangeRequest = parse_input(input)?;
            let (tasks, projects, cursor) = futures::try_join!(
                deps.tasks.list_range(ListTasksRangeInput {
                    user_id: user_id.clone(),
                    from: request.from,
                    to: request.to,
                }),
                deps.projects.list_by_user(&user_id),
                deps.tasks.sync_cursor(&user_id),
            )?;
            output(TaskRangeResponse { tasks, projects, cursor })
        }

        // 태스크 상세 화면 — 단건 조회
        "byId" => {
            let request: TaskIdRequest = parse_input(input)?;
            let task = deps
                .tasks
                .find_by_id(&TaskRefInput { user_id, id: request.id })
                .await?
                .ok_or_else(task_not_found)?;
            output(TaskResponse { task })
        }

        "create" => {
            let request: CreateTaskRequest = parse_input(input)?;
            if let Some(project_id) = &request.project_id {
                let project = deps
                    .projects
                    .find_by_id(&TaskRefInput { user_id: user_id.clone(), id: project_id.clone() })
                    .await?;
                if project.is_none() {
                    return Err(project_not_found());
                }
            }
            let task = deps
                .tasks
                .create(CreateTaskInput {
                    user_id: user_id.clone(),
                    title: request.title,
                    project_id: request.project_id,
                    date: request.date,
                    start_at: request.start_at,
                    duration_min: request.duration_min,
                })
                .await?;
            deps.sync.notify(&user_id);
            output(TaskResponse { task })
        }

        // 의도 기반 부분 업데이트 — patch의 필드만 최신 행에 적용한다.
        // 같은 필드의 경합은 나중 의도가 이기고(LWW), 다른 필드끼리는 충돌하지 않는다.
        "update" => {
            let request: UpdateTaskRequest = parse_input(input)?;
            if let Patch::Set(Some(project_id)) = &request.patch.project_id {
                let project = deps
                    .projects
                    .find_by_id(&TaskRefInput { user_id: user_id.clone(), id: project_id.clone() })
                    .await?;
                if project.is_none() {
                    return Err(project_not_found());
                }
            }
            let task = deps
                .tasks
                .update(UpdateTaskInput {
                    user_id: user_id.clone(),
                    id: request.id,
                    patch: request.patch,
                })
                .await?
                .ok_or_else(task_not_found)?;
            deps.sync.notify(&user_id);
            output(TaskResponse { task })
        }

        // 소프트 삭제 — tombstone이 델타로 다른 기기에 전파된다
        "delete" => {
            let request: TaskIdRequest = parse_input(input)?;
            let removed = deps
                .tasks
                .remove(&TaskRefInput { user_id: user_id.clone(), id: request.id.clone() })
                .await?;
            if !removed {
                return Err(task_not_found());
            }
            deps.sync.notify(&user_id);
            output(DeletedTaskResponse { id: request.id })
        }

        // 델타 동기화 — 커서 이후의 변경 전부 (tombstone 포함)
        "changes" => {
            let request: SyncChangesRequest = parse_input(input)?;
            let (tasks, projects) = futures::try_join!(
                deps.tasks.changes_since(ListChangesInput {
                    user_id: user_id.clone(),
                    cursor: request.cursor,
                }),
                deps.projects.changes_since(ListChangesInput {
                    user_id: user_id.clone(),
                    cursor: request.cursor,
                }),
            )?;
            let cursor = tasks
                .iter()
                .map(|change| change.sync_seq)
                .chain(projects.iter().map(|change| change.sync_seq))
                .fold(request.cursor, i64::max);
            output(SyncChangesResponse { tasks, projects, cursor })
        }

        "createProject" => {
            let request: CreateProjectRequest = parse_input(input)?;
            let project = deps
                .projects
                .create(CreateProjectInput {
                    user_id: user_id.clone(),
                    name: request.name,
                    color: request.color,
                })
                .await?;
            deps.sync.notify(&user_id);
            output(ProjectResponse { project })
        }

        // 프로젝트 목록 화면 — 프로젝트 라벨에 진행률(완료/전체)을 붙여 내려준다
        "projects" => {
            let (projects, counts) = futures::try_join!(
                deps.projects.list_by_user(&user_id),
                deps.tasks.counts_by_project(&user_id),
            )?;
            let count_by_id: HashMap<&str, &_> =
                counts.iter().map(|count| (count.project_id.as_str(), count)).collect();
            output(ProjectListResponse {
                projects: projects
                    .into_iter()
                    .map(|project| {
                        let count = count_by_id.get(project.id.as_str());
                        ProjectSummary {
                            total_count: count.map(|count| count.total).unwrap_or(0),
                            done_count: count.map(|count| count.done).unwrap_or(0),
                            project,
                        }
                    })
                    .collect(),
            })
        }

        // 프로젝트 상세(보드) — 프로젝트 라벨과 그 프로젝트의 태스크 전부
        "project" => {
            let request: ProjectDetailRequest = parse_input(input)?;
            let project = deps
                .projects
                .find_by_id(&TaskRefInput {
                    user_id: user_id.clone(),
                    id: request.project_id.clone(),
                })
                .await?
                .ok_or_else(project_not_found)?;
            let tasks = deps
                .tasks
                .list_by_project(ListTasksByProjectInput {
                    user_id,
                    project_id: request.project_id,
                })
                .await?;
            output(ProjectDetailResponse { project, tasks })
        }

        _ => unreachable!("디스패처가 존재하는 프로시저만 넘긴다"),
    }
}
